Add alternatebiddercodes in Prebid request ext

pm-nilesh-chate opened the pull request:
w.r.t https://github.com/prebid/prebid-server/issues/2174#issuecomment-1190736031

This changes would make alternatebiddercodes from PBS config formally part of Prebid ORTB extensions.

- Incoming ORTB would contain:
```
ext.prebid.alternatebiddercodes.enabled: true/false
ext.prebid.alternatebiddercodes.bidders.BIDDER.enabled: true/false
ext.prebid.alternatebiddercodes.bidders.BIDDER.allowedbiddercodes: [array, of, strings]
```
- This would override the account-level config
- If the request doesn't specify these values, then PBS-core would merge the account config into these values
- Before sending to each adapter, PBS-core would need to remove entries that don't belong to that bidder. i.e. they would only receive ext.prebid.alternatebiddercodes.adapters.ME
- Everything else is up to the adapter. They can ignore or use at will.

Ex. s2s config on page and incoming PBS request
```
pbjs.setConfig({
    s2sConfig: {
        extPrebid: {
            "alternatebiddercodes": {
                "enabled": true,
                "bidders": {
                    "pubmatic": {
                        "enabled": true,
                        "allowedbiddercodes": [
                            "groupm"
                        ]
                    }
                }
            }
        }
    }
});
```

```
{
   ...
   "ext": {
       "prebid": {
           "alternatebiddercodes": {
                "enabled": true,
                "bidders": {
                    "pubmatic": {
                        "enabled": true,
                        "allowedbiddercodes": [
                            "groupm"
                        ]
                    }
                }
            }
       }
   }
}
```

More details at https://github.com/prebid/prebid-server/issues/2174

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -297,13 +297,27 @@ func getBidderParamsForBidder(bidderParamsInReqExt map[string]map[string]json.Ra
+	extCopy.Prebid.AlternateBidderCodes = nil
+
+	//fill request.ext.prebid.alternatebiddercode to have data only of the current bidder
+	if unpackedExt.Prebid.AlternateBidderCodes != nil && len(unpackedExt.Prebid.AlternateBidderCodes.Bidders) != 0 {
```
Do we want to pass `openrtb_ext.ExtAlternateBidderCodes{Enabled: true/false}` to the bidders that are not present in the list req.ext?.

Current changes do not pass it as I don't find any use case for the same.

VeronikaSolovei9 commented:
Code looks good, local testing looks good too.

pm-nilesh-chate commented on config/accounts.go:
```diff
@@ -18,20 +16,22 @@ const (
+	CookieSync              CookieSync            `mapstructure:"cookie_sync" json:"cookie_sync"`
+	Events                  Events                `mapstructure:"events" json:"events"` // Don't enable this feature. It is still under developmment - https://github.com/prebid/prebid-server/issues/1725
+	TruncateTargetAttribute *int                  `mapstructure:"truncate_target_attr" json:"truncate_target_attr"`
+	AlternateBidderCodes    *AlternateBidderCodes `mapstructure:"alternatebiddercodes" json:"alternatebiddercodes"`
```
Reverted the change. Yes, I did it to detect the presence of account config values but I guess we don't need it.

btw, we would need the AlternateBidderCodes in the request.ext to be pointer to detect its presence and to choose between request and account values.

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
 	return res
 }
+
+// getExtAlternateBidderCodes map config.account.alternatebiddercodes to ext.prebid.alternatebiddercodes
```
done

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
 }
+
+// getExtAlternateBidderCodes map config.account.alternatebiddercodes to ext.prebid.alternatebiddercodes
+func setExtAlternateBidderCodes(requestExt *openrtb_ext.ExtRequest, cfgABC *config.AlternateBidderCodes) {
```
done

pm-nilesh-chate commented on exchange/exchange.go:
```diff
@@ -494,7 +496,7 @@ func (e *exchange) getAllBids(
 	globalPrivacyControlHeader string,
 	headerDebugAllowed bool,
-	alternateBidderCodes config.AlternateBidderCodes,
+	alternateBidderCodes *openrtb_ext.ExtAlternateBidderCodes,
```
done

pm-nilesh-chate commented on exchange/exchange.go:
```diff
@@ -245,6 +245,8 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		}
 	}
 
+	setExtAlternateBidderCodes(requestExt, r.Account.AlternateBidderCodes)
```
sure, moved `setExtAlternateBidderCodes`'s code into `CleanOpenRTBRequests`'s `buildRequestExtForBidder`.

It looks messy now, too many scenarios, although tried to cover all them in unit tests.

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -831,3 +845,15 @@ func WrapJSONInData(data []byte) []byte {
+
+// getExtAlternateBidderCodes map config.account.alternatebiddercodes to ext.prebid.alternatebiddercodes
+func setExtAlternateBidderCodes(requestExt *openrtb_ext.ExtRequest, cfgABC *config.AlternateBidderCodes) {
+	if requestExt == nil || requestExt.Prebid.AlternateBidderCodes != nil {
```
missed this one. Updated the code in `buildRequestExtForBidder` and added unit test for the same.

pm-nilesh-chate commented on exchange/exchangetest/extra-bids.json:
```diff
@@ -40,6 +40,15 @@
             "includewinners": true,
             "includebidderkeys": true
+          },
+          "alternatebiddercodes": {
+            "enabled": true,
+            "bidders": {
+              "pubmatic": {
+                "enabled": true,
+                "allowedbiddercodes": ["groupm"]
+              }
+            }
```
Agreed, also I could not test end to end flow at one place where we could validate outgoing and incoming request as these json file based tests mock `requestBid` function which has alternatebiddercodes validation.

Hence, I wrote `TestOverrideConfigAlternateBidderCodesWithRequestValues` to cover the validation of alternatebiddercodes coming as response from adapters and added unit tests in all the possible functions that could change `request.ext.prebid. alternatebiddercodes` to cover the request to adapters part. Ex. `setExtAlternateBidderCodes` and `CleanOpenRTBRequests` (now clubbed).

pm-nilesh-chate replied:
added `expectRequest`

pm-nilesh-chate commented:
Hi @bsardo, I've address the comments. Could you review the same.

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -48,6 +49,7 @@ func cleanOpenRTBRequests(ctx context.Context,
 	gdprPermsBuilder gdpr.PermissionsBuilder,
 	tcf2ConfigBuilder gdpr.TCF2ConfigBuilder,
 	hostSChainNode *openrtb2.SupplyChainNode,
+	cfgABC config.AlternateBidderCodes,
```
done

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
+				Bidders: map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
+					bidder: bidderCodes,
+				},
+			}
```
done. Added similar test case for `buildRequestExtForBidder()`

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
+				return json.RawMessage(``), nil
+			}
+			alternateBidderCodes = nil
+		}
```
done

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -273,9 +276,36 @@ func getAuctionBidderRequests(auctionRequest AuctionRequest,
-	if len(requestExt) == 0 || requestExtParsed == nil {
-		return json.RawMessage(``), nil
+func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, requestExtParsed *openrtb_ext.ExtRequest, bidderParamsInReqExt map[string]json.RawMessage, cfgABC config.AlternateBidderCodes) (json.RawMessage, error) {
+	// Resolve alternatebiddercode for current bidder
```
done

pm-nilesh-chate commented:
@bsardo I've address the comments. Could you review the same.

Thank you for the critical inputs. A good learning for my future PRs :)

VeronikaSolovei9 reviewed (approved):
Code looks so much better and easier to understand now. LGTM

bsardo commented on exchange/utils_test.go:
```diff
@@ -3272,3 +3302,159 @@ func TestBuildExtData(t *testing.T) {
+			inExt:    nil,
+			inCfgABC: config.AlternateBidderCodes{Enabled: true},
+			wantExt: []json.RawMessage{
+				json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":null}}}`),
+				json.RawMessage(`{"prebid":{"alternatebiddercodes":{"enabled":true,"bidders":null}}}`),
```
I see a few instances where `prebid.alternatebiddercodes.bidders` is `null`. Perhaps this should be omitted if null, in which case we would add the `omitempty` tag to the `ExtAlternateBidderCodes.Bidders` field. Thoughts?

pm-nilesh-chate replied:
Added omit empty to all the alternatebiddercodes fields.

The only downside I see here is that the SSPs would not know whether these fields were omitted by the user or by the PBS. One will have to refer the incoming payload to PBS to verify the same. I don't see any harm though.

pm-nilesh-chate commented on exchange/utils.go:
```diff
@@ -302,14 +313,42 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
+		return &openrtb_ext.ExtAlternateBidderCodes{
+			Enabled: reqABC.Enabled,
+			Bidders: map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
+				bidder: reqABC.Bidders[bidder],
```
done. Also added a test case for this scenario

pm-nilesh-chate commented on exchange/exchange.go:
```diff
@@ -286,7 +286,14 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		liveAdapters = listBiddersWithRequests(bidderRequests)
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, r.Account.AlternateBidderCodes, requestExt.Prebid.Experiment)
+
+		//This will be used validate bids
+		alternateBidderCodes := openrtb_ext.ExtAlternateBidderCodes(r.Account.AlternateBidderCodes)
+		if requestExt != nil && requestExt.Prebid.AlternateBidderCodes != nil {
+			alternateBidderCodes = *requestExt.Prebid.AlternateBidderCodes
+		}
```
Added this to make sure PBS-Core data is used to validate the adapter's response instead of relying on the data from adapter's request which could be altered by the adapter code.

pm-nilesh-chate commented:
Addressed the review comments. Had to rebase and force push to fix the master merge test failures.

@bsardo Could you please review the latest changes.

Thanks

abhinavsinha001 commented:
What is the holdup here can we get these changes merged? 
We want to start using this functionality.

bsardo commented:
@abhinavsinha001 I'm waiting for a response on this https://github.com/prebid/prebid-server/issues/2174#issuecomment-1242045987

abhinavsinha001 commented:
Thanks @bsardo for pointing to the comment.
@bretg can you please help unblock here.

bsardo commented on exchange/utils.go:
```diff
@@ -302,14 +313,45 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
 	return json.Marshal(extMap)
 }
 
+func buildRequestExtAlternateBidderCodes(bidder string, accABC config.AlternateBidderCodes, reqABC *openrtb_ext.ExtAlternateBidderCodes) *openrtb_ext.ExtAlternateBidderCodes {
```
After looking more closely at the requirements, I believe we should aim for the following behavior:
- If `alternatebiddercodes` is not defined in the account or on the request, the field should be omitted from all bidder requests.
- If `alternatebiddercodes` is defined either in the account or on the request but its `enabled` is not set, the `alternatebiddercodes` field should be added to the bidder request with the default value of false.
- The `alternatebiddercodes` object set on a bidder request should only contain its own bidder object in the `bidders` field and it should match the bidder object in the account/request it was sourced from. Note an undefined map or slice in the original request or account will appear as `null` in a bidder request (we will need to get rid of `omitempty`), so that we can distinguish between an empty array/map and an undefined array/map.

Given the above, I'd like to suggest a couple of things here:
1) In `config/accounts.go`, can we make the field `AlternateBidderCodes` in the `Account` struct of type `*AlternateBidderCodes`?
2) Let's get rid of the `if` block on line 348 that sets `alternateBidderCodes = nil`.
3) I think this function ultimately looks like this:
```
        if reqABC != nil {
		alternateBidderCodes := &openrtb_ext.ExtAlternateBidderCodes{
			Enabled: reqABC.Enabled,
		}
		if bidderCodes, ok := reqABC.Bidders[bidder]; ok {
			alternateBidderCodes.Bidders = map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
				bidder: bidderCodes,
			}
		}
		return alternateBidderCodes
	}

	if accABC != nil {
		alternateBidderCodes := &openrtb_ext.ExtAlternateBidderCodes{
			Enabled: accABC.Enabled,
		}
		if bidderCodes, ok := accABC.Bidders[bidder]; ok {
			alternateBidderCodes.Bidders = map[string]openrtb_ext.ExtAdapterAlternateBidderCodes{
				bidder: bidderCodes,
			}
		}
		return alternateBidderCodes
	}
	return nil
```

pm-nilesh-chate replied:
done. made `accABC` pointer to detect its presence.

bsardo commented on exchange/exchange.go:
```diff
@@ -286,7 +286,14 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
-		adapterBids, adapterExtra, anyBidsReturned = e.getAllBids(auctionCtx, bidderRequests, bidAdjustmentFactors, conversions, accountDebugAllow, r.GlobalPrivacyControlHeader, debugLog.DebugOverride, r.Account.AlternateBidderCodes, requestExt.Prebid.Experiment)
+
+		//This will be used validate bids
+		alternateBidderCodes := openrtb_ext.ExtAlternateBidderCodes(r.Account.AlternateBidderCodes)
```
See my related comment in `exchange/utils.go` that asks for the account field to be converted to a pointer.
Should this be something like this now:
```
var alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes
if r.Account.AlternateBidderCodes != nil {
    alternateBidderCodes = openrtb_ext.ExtAlternateBidderCodes(*r.Account.AlternateBidderCodes)
}
```

pm-nilesh-chate replied:
done

bsardo commented on exchange/utils.go:
```diff
@@ -302,14 +313,45 @@ func buildRequestExtForBidder(bidder string, requestExt json.RawMessage, request
+		}
+	}
+
+	if !alternateBidderCodes.Enabled && alternateBidderCodes.Bidders == nil {
```
I'm wondering now if we should delete this if block. If we remove it, `ext.prebid.alternatebiddercodes` in the bidder request will contain exactly what's pertinent to the bidder in the account config with the lone exception being that `enabled` flags will be set to the default value false if not defined. This is the same behavior as if `ext.prebid.alternatebiddercodes` were to be populated on the bidder request with information pulled from the original request, assuming we remove the couple of `omitempty` tags I pointed out in my other comments.

pm-nilesh-chate replied:
agreed. removed this code.

bsardo reviewed (commented):
Thanks for your patience @pm-nilesh-chate. Please see my comments. I ran a bunch of manual tests with my suggested changes to make sure they were in line with my requirements analysis outlined in my comment in `exchange/utils.go`. Let me know if you want to discuss.

pm-nilesh-chate commented on openrtb_ext/alternatebiddercodes.go:
```diff
@@ -0,0 +1,53 @@
+
+type ExtAdapterAlternateBidderCodes struct {
+	Enabled            bool     `mapstructure:"enabled" json:"enabled"`
+	AllowedBidderCodes []string `mapstructure:"allowedbiddercodes" json:"allowedbiddercodes,omitempty"`
```
done, removed `omitempty` tag.

pm-nilesh-chate commented on openrtb_ext/alternatebiddercodes.go:
```diff
@@ -0,0 +1,53 @@
+// ExtAlternateBidderCodes defines list of alternate bidder codes allowed by adatpers. This overrides host level configs.
+type ExtAlternateBidderCodes struct {
+	Enabled bool                                      `mapstructure:"enabled" json:"enabled"`
+	Bidders map[string]ExtAdapterAlternateBidderCodes `mapstructure:"bidders" json:"bidders,omitempty"`
```
done, removed `omitempty` tag.

pm-nilesh-chate commented:
Hi @bsardo, I've addressed the comments. Please review the same. Fyi, last commit (master merge) is to address Trivy's check.

bsardo reviewed (dismissed):
LGTM

SyntaxNode commented on openrtb_ext/alternatebiddercodes.go:
```diff
@@ -0,0 +1,53 @@
+
+func (bidderCodes *ExtAlternateBidderCodes) IsValidBidderCode(bidder, alternateBidder string) (bool, error) {
+	const ErrAlternateBidderNotDefined = "alternateBidderCodes not defined for adapter %q, rejecting bids for %q"
+	const ErrAlternateBidderDisabled = "alternateBidderCodes disabled for %q, rejecting bids for %q"
```
The naming convention `errSomeName` is usually for when the value is an error type. This is an error message and the separation of the fields from the Errorf calls is a bit hard to follow IMHO. Consider extracting each to their own simple builder method, like:

```
func alternateBidderNotDefinedError(bidder, alternateBidder string) error {
  return fmt.Errorf("alternateBidderCodes not defined for adapter %q, rejecting bids for %q", bidder, alternateBidder)
}
```

pm-nilesh-chate replied:
done

bsardo commented on exchange/exchange.go:
```diff
@@ -286,7 +286,16 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
+		if requestExt != nil && requestExt.Prebid.AlternateBidderCodes != nil {
+			alternateBidderCodes = *requestExt.Prebid.AlternateBidderCodes
+		} else if r.Account.AlternateBidderCodes != nil {
+			alternateBidderCodes = openrtb_ext.ExtAlternateBidderCodes(*r.Account.AlternateBidderCodes)
```
I think you can remove this cast now since `r.Account.AlternateBidderCodes` is already of type `ExtAlternateBidderCodes`.

VeronikaSolovei9 reviewed (approved):
LGTM
